core/embed/rust/src/ui/display.rs

```diff
@@ -105,6 +151,341 @@ pub fn rect_fill_rounded1(r: Rect, fg_color: Color, bg_color: Color) {
+                    || (inverted && !(is_past_start && is_before_end))
+                {
+                    let p_b = p - r.top_left();
+                    let c =
+                        rect_rounded2_get_pixel(p_b, r.width(), r.height(), colortable, false, 2);
+                    pixeldata(c);
+                } else {
+                    pixeldata(bg_color);
+                }
+            }
+        }
+    }
+
+    pixeldata_dirty();
+}
+
+/// Gets a color of a pixel on `p` coordinates of rounded rectangle with corner
+/// radius 2
+pub fn rect_rounded2_get_pixel(
+    p: Offset,
+    w: i32,
+    h: i32,
+    colortable: [Color; 16],
+    fill: bool,
+    line_width: i32,
+) -> Color {
+    let border = (p.x >= 0 && p.x < line_width)
+        || ((p.x >= w - line_width) && p.x <= (w - 1))
+        || (p.y >= 0 && p.y < line_width)
+        || ((p.y >= h - line_width) && p.y <= (h - 1));
+
+    let corner_lim = 2 * line_width;
+    let corner_inner = line_width;
+
+    let corner_all = ((p.x > w - (corner_lim + 1)) || p.x < corner_lim)
+        && (p.y < corner_lim || p.y > h - (corner_lim + 1));
+
+    let corner = corner_all
+        && (p.y >= corner_inner)
+        && (p.x >= corner_inner)
+        && (p.y <= h - (corner_inner + 1))
+        && (p.x <= w - (corner_inner + 1));
+
+    let corner_out = corner_all && !corner;
+
+    if (border || corner || fill) && !corner_out {
+        colortable[15]
+    } else {
+        colortable[0]
+    }
+}
+
+/// Draws a rounded rectangle with corner radius 2, partially filled
+/// according to `fill_from` and `fill_to` arguments.
+/// Optionally draws a text inside the rectangle and adjusts its color to match
+/// the fill. The coordinates of the text are specified in the TextOverlay
+/// struct.
+pub fn bar_with_text_and_fill(
+    area: Rect,
+    overlay: Option<TextOverlay>,
+    fg_color: Color,
+    bg_color: Color,
+    fill_from: i32,
+    fill_to: i32,
+) {
+    let r = area.translate(get_offset());
+    let clamped = r.clamp(constant::screen());
+    let colortable = get_color_table(fg_color, bg_color);
+
+    set_window(clamped);
+
+    for y_c in clamped.y0..clamped.y1 {
+        for x_c in clamped.x0..clamped.x1 {
+            let p = Point::new(x_c, y_c);
+            let r_offset = p - r.top_left();
+
+            let filled = (r_offset.x >= fill_from
+                && fill_from >= 0
+                && (r_offset.x <= fill_to || fill_to < fill_from))
+                || (r_offset.x < fill_to && fill_to >= 0);
+
+            let underlying_color =
+                rect_rounded2_get_pixel(r_offset, r.width(), r.height(), colortable, filled, 1);
+
+            let mut overlay_color = None;
+            if let Some(o) = overlay {
+                overlay_color = o.get_pixel(None, p);
+            }
+
+            let mut final_color = underlying_color;
+
+            if let Some(overlay) = overlay_color {
+                if overlay == fg_color {
```

Comment:
```
let overlay_color = overlay.map_or(underlying_color, |o| o.get_pixel(underlying_color, p))
let final_color = if (overlay_color == fg_color) {
    underlying_color.negative()
} else {
    underlying_color
}
```

Reply:
done https://github.com/trezor/trezor-firmware/commit/ce97b6452904d229e17dc2e672562804e24a71f2

(differs a bit as the `get_pixel` now accepts the text fg color from here)